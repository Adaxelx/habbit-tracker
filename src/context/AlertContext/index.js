export { default, useAlertContext } from './AlertContext';
