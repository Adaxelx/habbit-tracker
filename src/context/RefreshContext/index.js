export { default, useRefreshContext } from './RefreshContext';
