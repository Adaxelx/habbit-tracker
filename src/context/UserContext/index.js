export { default, useUserContext, UserContext } from './UserContext';
