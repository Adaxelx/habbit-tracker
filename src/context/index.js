export { default as ThemeProvider, useThemeContext } from './ThemeContext';
export { default as UserProvider, useUserContext, UserContext } from './UserContext';
export { default as AlertProvider, useAlertContext, AlertContext } from './AlertContext';
export { default as RefreshProvider, useRefreshContext } from './RefreshContext';
