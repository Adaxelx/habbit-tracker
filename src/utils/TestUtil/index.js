export { default } from './TestUtil.ts';
