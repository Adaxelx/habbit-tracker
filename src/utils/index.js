export { default as TestUtil } from './TestUtil';
export { default as RenderWithRouter } from './RenderWithRouter';
export * from './validators';
export * from './types.d';
