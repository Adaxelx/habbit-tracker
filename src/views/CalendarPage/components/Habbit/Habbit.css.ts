import styled, { css } from 'styled-components';

export const StyledContainer = styled.div`
  margin-bottom: ${({ theme }) => theme.margin.xs};
`;

interface LabelProps {
  color: string;
}

export const StyledLabel = styled.small<LabelProps>`
  background-color: ${({ color }) => color};
  padding: ${({ theme }) => theme.margin.s};
  width: auto;
  border-radius: 20px;
`;

interface HabbitProps {
  checked?: boolean;
}

const checkedStyles = css`
  text-decoration: line-through;
  color: gray;
`;

export const StyledHabbit = styled.div<HabbitProps>`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  padding: 0 ${({ theme }) => theme.margin.s};
  border-left: 1px solid ${({ theme }) => theme.colors.tile.border};
  margin: 0 ${({ theme }) => theme.margin.s};
  margin-top: ${({ theme }) => theme.margin.xs};
  overflow-y: scroll;
  &:hover {
    cursor: pointer;
  }
  ${({ checked }) => checked && checkedStyles};
`;

export const StyledTitle = styled.h2`
  font-size: ${({ theme }) => theme.font.sizes.ml};
  margin-bottom: ${({ theme }) => theme.margin.xs};
  margin-top: 0;
`;

export const StyledTime = styled.p`
  color: gray;
  margin-bottom: ${({ theme }) => theme.margin.xs};
`;

export const StyledDescription = styled.p`
  text-align: justify;
`;
