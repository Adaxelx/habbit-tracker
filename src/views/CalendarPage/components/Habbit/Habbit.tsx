import React, { useRef } from 'react';
import { Button } from 'components';
import { Event, AlertTypes } from 'utils';
import { useAlertContext, useUserContext } from 'context';
import { deleteEvent } from 'views/CalendarPage/CalendarPage.api';
import {
  StyledHabbit,
  StyledTime,
  StyledDescription,
  StyledTitle,
  StyledContainer,
  StyledLabel,
} from './Habbit.css';

const { SUCCESS } = AlertTypes;

const Habbit = ({ habbit, handleRefresh }: { habbit: Event; handleRefresh: Function }) => {
  const { title, timeEnd, timeStart, description, label, _id } = habbit;

  const alertC = useRef(useAlertContext());
  const { token } = useUserContext();

  const handleDelete = async () => {
    try {
      await deleteEvent(token, _id);
      alertC.current.showAlert('Succesfuly deleted habbit.', SUCCESS);
      handleRefresh();
    } catch (err) {
      alertC.current.showAlert(err.message);
    }
  };

  return (
    <StyledContainer>
      <Button size="s" mx="0.75rem" noMaxWidth data-testid="edit">
        Edit habbit
      </Button>
      <Button size="s" close noMaxWidth data-testid="delete" onClick={handleDelete}>
        X
      </Button>
      <StyledHabbit>
        {label && <StyledLabel color={label.color}>{label.title}</StyledLabel>}
        <StyledTitle>{title}</StyledTitle>
        <StyledTime>{`${timeStart}-${timeEnd}`}</StyledTime>
        <StyledDescription>{description}</StyledDescription>
      </StyledHabbit>
    </StyledContainer>
  );
};

export default Habbit;
