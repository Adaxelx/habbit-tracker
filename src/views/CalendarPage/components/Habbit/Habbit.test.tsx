import React from 'react';
import 'jest-styled-components';
import { TestUtil, HabbitProps, AlertTypes } from 'utils';
import Habbit from './Habbit';
import { AlertContext } from 'context';
import { waitFor } from '@testing-library/dom';

describe('Habbit', () => {
  let util: TestUtil;
  const alertC = { showAlert: jest.fn() };
  let props: HabbitProps;
  let fail: boolean;
  const mockedFetch = (input: string, init: { method: string }) => {
    if (fail) {
      return Promise.resolve({
        status: 500,
        ok: false,
        json: () => Promise.resolve({ message: 'error' }),
      });
    }
    switch (init.method) {
      case 'PATCH':
        return Promise.resolve({
          status: 200,
          ok: true,
          json: () => Promise.resolve({ message: 'Successfully state change.' }),
        });
      case 'DELETE':
        return Promise.resolve({
          status: 204,
          ok: true,
        });
      default:
        return Promise.resolve({});
    }
  };

  beforeAll(() => {
    jest.spyOn(window, 'fetch');
  });

  beforeEach(() => {
    jest.clearAllMocks();
    fail = false;
    props = {
      habbit: {
        _id: '123',
        userId: '123',
        title: 'habbit',
        description: 'desc',
        daysOfWeek: [0, 2, 4, 6],
        dateStart: '2021-06-06',
        dateEnd: '2021-09-07',
        timeStart: '06:06',
        timeEnd: '07:07',
        label: { _id: '123', userId: '123', title: 'label', color: '#676767' },
      },
      labels: [{ _id: '123', userId: '123', title: 'label', color: '#676767' }],
      checked: false,
      day: [2021, 6, 6],
    };
    util = new TestUtil(
      (
        <AlertContext.Provider value={alertC}>
          <Habbit {...props} />
        </AlertContext.Provider>
      ),
    );
    window.fetch.mockImplementation(mockedFetch);
  });

  it('should match snapshot', () => {
    expect(util.render.asFragment()).toMatchSnapshot();
  });

  it('should delete habbit', async () => {
    util.click('delete');

    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith(
      'Successfully deleted habbit.',
      AlertTypes.SUCCESS,
    );
  });

  it('should not delete habbit if api fails', async () => {
    fail = true;
    util.click('delete');

    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('error');
  });

  it('should check habbit', async () => {
    util.click('habbit123');

    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('Successful state change.', AlertTypes.SUCCESS);
  });

  it('should not check habbit if api fails', async () => {
    fail = true;
    util.click('habbit123');

    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('error');
  });
});
