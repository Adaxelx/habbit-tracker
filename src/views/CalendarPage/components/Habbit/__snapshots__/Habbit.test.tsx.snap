// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Habbit should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  margin-bottom: 0.5rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  padding: 0 0.75rem;
  border-left: 1px solid #77CBFF;
  margin: 0 0.75rem;
  margin-top: 0.5rem;
}

.c3:hover {
  cursor: pointer;
}

.c4 {
  font-size: 2rem;
  margin-bottom: 0.5rem;
  margin-top: 0;
}

.c5 {
  color: gray;
  margin-bottom: 0.5rem;
}

.c6 {
  text-align: justify;
}

.c1 {
  color: #fff;
  background-color: #FFC136;
  border: 1px solid #707070;
  padding: 0.75rem;
  font-size: 1rem;
  min-width: none;
  display: inline-block;
  text-align: center;
  margin-top: 0;
  margin-left: 0.75rem;
  margin-right: 0.75rem;
}

.c1:hover {
  cursor: pointer;
}

.c1:disabled {
  -webkit-filter: grayscale(1);
  filter: grayscale(1);
}

.c2 {
  color: #000000;
  background-color: #FF9191;
  border: 1px solid #707070;
  padding: 0.75rem;
  font-size: 1rem;
  min-width: none;
  display: inline-block;
  text-align: center;
  margin-top: 0;
  margin-left: 0;
  margin-right: 0;
}

.c2:hover {
  cursor: pointer;
}

.c2:disabled {
  -webkit-filter: grayscale(1);
  filter: grayscale(1);
}

<div
    class="c0"
  >
    <button
      class="c1"
      data-testid="edit"
    >
      Edit habbit
    </button>
    <button
      class="c2"
      data-testid="delete"
    >
      X
    </button>
    <div
      class="c3"
    >
      <h2
        class="c4"
      >
        Title
      </h2>
      <p
        class="c5"
      >
        12:00-24:00
      </p>
      <p
        class="c6"
      >
        test
      </p>
    </div>
  </div>
</DocumentFragment>
`;
